File: 0x10 基本数据结构/0x16 Trie/AcWing142.py
N = 1000006
trie, tot, s = [[0]*27 for _ in range(N)], 1, ""

def add():
    global tot, s
    len_, p = len(s), 1
    for i in range(len_):
        k = ord(s[i]) - ord('a') + 1
        if not trie[p][k]: tot += 1; trie[p][k] = tot
        p = trie[p][k]
    trie[p][0] += 1
    
def get():
    global s
    ans, len_, p = 0, len(s), 1
    for i in range(len_):
        p = trie[p][ord(s[i])-ord('a')+1]
        ans += trie[p][0]
    print(ans)
    
def main():
    global s
    n, m = [int(x) for x in input().split()]
    for _ in range(n):
        s = input()
        add()
    for _ in range(m):
        s = input()
        get()

main()

File: 0x10 基本数据结构/0x16 Trie/AcWing142.cpp
#include <cstring>
#include <iostream>
using namespace std;
const int N = 1000006;
int trie[N][27], tot = 1;
char s[N];

void add() {
  int len = strlen(s), p = 1;
  for (int i=0; i<len; i++) {
    int k = s[i] - 'a' + 1;
    if (!trie[p][k]) trie[p][k] = ++tot;
    p = trie[p][k];
  }
  ++trie[p][0]; // 为处理插入重复字符串的情况，不能只做结尾标记
}

void get() {
  int ans = 0, len = strlen(s), p = 1;
  for (int i=0; i<len; i++) {
    p = trie[p][s[i]-'a'+1];
    ans += trie[p][0];
  }
  cout << ans << endl;
}

int main() {
  memset(trie, 0, sizeof(trie));
  int n, m;
  cin >> n >> m;
  while (n--) {
    scanf("%s", s);
    add();
  }
  while (m--) {
    scanf("%s", s);
    get();
  }
  return 0;
}